// urequire/build/code/codeUtils/toCode.js
// Generated by uRequire v0.7.0-beta.23 target: 'lib' template: 'nodejs'

var _ = require('lodash'),
    _B = require('uberscore'),
    UError = require('../utils/UError');

var l = new _B.Logger('uRequire/codeUtils/toCode');

var dfb, escodegen, toAST, toCode;
escodegen = require("escodegen");
dfb = new _B.DeepDefaultsBlender();
toAST = require("./toAST");
module.exports = toCode = function (astCode, options) {
  var ast, err, error;
  if (options == null) {
    options = toCode.options;
  }
  if (_.isEmpty(astCode)) {
    return "";
  }
  if (options !== toCode.options) {
    dfb.blend(options, toCode.options);
  }
  ast = toAST(astCode, options.type);
  try {
    return escodegen.generate(ast, options);
  } catch (error) {
    err = error;
    throw new UError("toCode: Error generating code from AST with escodegen: " + err + ".", { nested: err });
  }
};
toCode.options = {
  format: {
    indent: {
      style: "  ",
      base: 0
    },
    json: false,
    renumber: false,
    hexadecimal: false,
    quotes: "double",
    escapeless: true,
    compact: false,
    parentheses: true,
    semicolons: true
  }
};

;
