// urequire/build/code/config/MasterDefaultsConfig.js
// Generated by uRequire v0.7.0-beta.23 target: 'lib' template: 'nodejs'

var _ = require('lodash'),
    _B = require('uberscore'),
    UError = require('../utils/UError');

var l = new _B.Logger('uRequire/config/MasterDefaultsConfig.coffee');

var MasterDefaultsConfig;
module.exports = MasterDefaultsConfig = {
  bundle: {
    path: void 0,
    filez: void 0,
    copy: [],
    name: void 0,
    main: void 0,
    dependencies: {
      imports: {},
      rootExports: {},
      replace: {},
      node: [
        "fs",
        "events",
        "util",
        "http",
        "path",
        "child_process",
        "events",
        "crypto",
        "string_decoder",
        "timers",
        "tls",
        "domain",
        "buffer",
        "net",
        "dgram",
        "dns",
        "stream",
        "https",
        "querystring",
        "punycode",
        "readline",
        "url",
        "repl",
        "vm",
        "assert",
        "tty",
        "zlib",
        "os",
        "cluster",
        "console",
        "freelist",
        "sys",
        "constants"
      ],
      locals: {},
      paths: {
        useCache: true,
        override: void 0,
        bower: void 0,
        npm: void 0
      },
      shim: true,
      depsVars: {},
      _knownDepsVars: {
        chai: "chai",
        mocha: "mocha",
        lodash: "_",
        underscore: "_",
        jquery: [
          "$",
          "jQuery"
        ],
        backbone: "Backbone",
        knockout: [
          "ko",
          "Knockout"
        ]
      }
    },
    resources: require("./ResourceConverters").defaultResourceConverters,
    commonCode: void 0,
    webRootMap: "."
  },
  build: {
    dstPath: void 0,
    target: void 0,
    forceOverwriteSources: false,
    template: "UMDplain",
    runtimeInfo: true,
    bare: false,
    useStrict: void 0,
    globalWindow: false,
    injectExportsModule: true,
    rootExports: {
      ignore: false,
      runtimes: ["script"],
      noConflict: true
    },
    scanAllow: false,
    allNodeRequires: false,
    dummyParams: false,
    noLoaderUMD: false,
    warnNoLoaderUMD: true,
    watch: false,
    verbose: false,
    debugLevel: 0,
    "continue": false,
    optimize: false,
    _optimizers: [
      "uglify2",
      "uglify"
    ],
    out: void 0,
    afterBuild: [],
    clean: true,
    deleteErrored: true,
    rjs: void 0
  }
};

;
