// urequire/build/code/config/ResourceConverters.js
// Generated by uRequire v0.7.0-beta.23 target: 'lib' template: 'nodejs'

var _ = require('lodash'),
    _B = require('uberscore'),
    UError = require('../utils/UError');

var l = new _B.Logger('uRequire/config/ResourceConverters.coffee');

var ResourceConverter, defaultResourceConverters, wrapCoffeeDefineCommonJS;
defaultResourceConverters = [
  {
    name: "$javascript",
    descr: "Dummy js converter, justs marks `.js` files as `Module`s.",
    filez: [
      "**/*.js",
      /.*\.(javascript)$/
    ],
    convert: function (modyle) {
      return modyle.converted;
    },
    convFilename: function (srcFilename) {
      return require("upath").changeExt(srcFilename, "js");
    },
    type: "module"
  },
  "coffee-script",
  "livescript",
  [
    "coco",
    { bare: false }
  ]
];
exports.defaultResourceConverters = defaultResourceConverters;
ResourceConverter = require("./ResourceConverter");
defaultResourceConverters.unshift(wrapCoffeeDefineCommonJS = new ResourceConverter([
  "wrapCoffeeDefineCommonJS",
  [
    "**/*.coffee",
    "**/*.co",
    "**/*.ls",
    "**.iced"
  ],
  function (r) {
    var i, len, line, lines;
    lines = r.converted.split("\n");
    r.converted = "define ->\n";
    for (i = 0, len = lines.length; i < len; i++) {
      line = lines[i];
      if (line) {
        r.converted += "  " + line + "\n";
      }
    }
    return r.converted += "  return module.exports";
  }
]));
wrapCoffeeDefineCommonJS.enabled = false;

;
